rasa/core/lock_store.py

```diff
@@ -0,0 +1,227 @@
+    def find_lock_store(store=None) -> "LockStore":
+        if store is None or store.type is None or store.type == "in_memory":
+            lock_store = InMemoryLockStore()
+        elif store.type == "redis":
+            lock_store = RedisLockStore(host=store.url, **store.kwargs)
+        else:
+            raise ValueError(
+                "Cannot create `LockStore` of type '{}'. One of the following "
+                "`LockStore` types need to be specified: {}."
+                "".format(store.type, ", ".join(ACCEPTED_LOCK_STORES))
+            )
+
+        logger.debug(
+            "Connected to lock store {}.".format(lock_store.__class__.__name__)
+        )
+
+        return lock_store
+
+    def create_lock(self, conversation_id: Text) -> TicketLock:
+        """Create and save a new `TicketLock` for `conversation_id`."""
+
+        lock = TicketLock(conversation_id)
+        self.save_lock(lock)
+        return lock
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Fetch lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Delete lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit `lock` to storage."""
+
+        raise NotImplementedError
+
+    def issue_ticket(
+        self, conversation_id: Text, lifetime: Union[float, int] = None
+    ) -> int:
+        """Issue new ticket for lock associated with `conversation_id`.
+
+        Creates a new lock if none is found.
+        """
+
+        lock = self.get_or_create_lock(conversation_id)
+
+        lifetime = lifetime if lifetime is not None else self.lifetime
+        ticket = lock.issue_ticket(lifetime)
+
+        self.save_lock(lock)
+        return ticket
+
+    @asynccontextmanager
+    @async_generator
+    async def lock(
+        self, conversation_id: Text, attempts: int = 60, wait: Union[int, float] = 1
+    ) -> None:
+        """Acquire lock for `conversation_id` with `ticket`.
+
+        Perform `attempts` with a wait of `wait` seconds between them before
+        raising a `LockError`.
+        """
+
+        ticket = self.issue_ticket(conversation_id)
+
+        try:
+            # have to use async_generator.yield_() for py 3.5 compatibility
+            await yield_(
+                await self._acquire_lock(conversation_id, ticket, attempts, wait)
+            )
+        finally:
+            self.cleanup(conversation_id, ticket)
+
+    async def _acquire_lock(
+        self, conversation_id: Text, ticket: int, attempts: int, wait: Union[int, float]
+    ) -> TicketLock:
+
+        while attempts > 0:
+            # fetch lock in every iteration because lock might no longer exist
+            lock = self.get_lock(conversation_id)
+
+            # exit loop if lock does not exist anymore (expired)
+            if not lock:
```

Comment:
should we put this in a function `_is_expired(lock)` ?

Reply:
I'm not sure about that, it would really just be
```python
def _is_expired(lock):
    if not lock:
        return True
```
plus additional code 
```python
if self._is_expired(lock):
    break
```

seems very unnecessary to me